docs/developer/architecture/development-plugin-saved-objects.asciidoc

```diff
@@ -202,4 +202,226 @@ up to date. If a visualization `id` was directly stored in
+[source,typescript]
+----
+import { SavedObjectsType } from 'src/core/server';
+
+export const dashboardVisualization: SavedObjectsType = {
+  name: 'dashboard_visualization', // <1>
+  /** ... */
+  migrations: {
+    // Takes a pre 1.1.0 doc, and converts it to 1.1.0
+    '1.1.0': (doc: DashboardVisualizationPre110): DashboardVisualization110 => { // <1>
+      return {
+        ...doc, // <2>
+        attributes: {
+          ...doc.attributes,
+          title: `${doc.attributes.title} - ${doc.attributes.subtitle}`
+        }
+      }
+    },
+
+    // Takes a 1.1.0 doc, and converts it to 1.4.0
+    '1.4.0': (doc: DashboardVisualization110): DashboardVisualization140 => { // <3>
+      doc.attributes.panels = doc.attributes.panels.map(panel => {
+        panel.id = uuid.v4();
+        return panel;
+      });
+      return doc;
+    },
+  },
+};
+----
+<1> It is useful to define an interface for each version of the schema. This
+allows TypeScript to ensure that you are properly handling the input and output
+types correctly as the schema evolves.
+<2> Returning a shallow copy is necessary to avoid type errors when using
+different types for the input and output shape.
+<3> Migrations do not have to be defined for every version, only those in which
+the schema needs to change.
+
+Migrations should be written defensively. If a document is encountered that is
+not in the expected shape, migrations are encouraged to throw an exception to
+abort the upgrade.
+
+===== Nested Migrations
+
+In some cases, objects may contain state that is functionally "owned" by other
+plugins. An example is a dashboard that may contain state owned by specific
+types of embeddables. In this case, the dashboard migrations should delegate
+migrating this state to their functional owner, the individual embeddable types,
+in order to compose a single migration function that handles all nested state.
+
+How the migration of the nested object is surfaced to the containing object is
+up to plugin authors to define. In general, we encourage that registries that
+expose state that may be persisted elsewhere include migrations as part of the
+interface that is registered. This allows consumers of the registry items to
+utilize these migrations on the nested pieces of state inside of the root
+migration.
+
+To demonstrate this, let's imagine we have:
+- A chart registry plugin which allows plugins to register multiple types of charts
+- A bar chart plugin that implements and registers a type of chart
+- A chart list plugin that contains a list of charts to display. This plugin
+persists the state of the underlying charts and allows those charts to define
+migrations for that state.
+
+The example code here is length, but necessary to demonstrate how this different
+pieces fit together.
+
+src/plugins/dashboard/server/saved_objects/dashboard.ts
+[source,typescript]
+----
+type ChartMigrationFn<ChartState extends object = object> = (chartDoc: object) => ChartState;
```

Comment:
that's basically the same as ` type ChartMigrationFn = (input: Record<string, any>) => Record<string, any>`, we cannot guarantee the output type for *every* migration